json_rpc/streamconnection.nim

```diff
@@ -0,0 +1,148 @@
+  ./json_rpc/[server, client]
+
+export jsonmarshal, router, server
+
+type
+  StreamClient* = ref object of RpcClient
+    output*: AsyncOutputStream
+  StreamConnection* = ref object of RpcServer
+    output*: AsyncOutputStream
+    client*: StreamClient
+  Handler[T, U] = proc(input: T): Future[U] {.gcsafe, raises: [Defect, CatchableError, Exception].}
+  HandlerWithId[T, U] = proc(input: T, id: int): Future[U] {.gcsafe, raises: [Defect, CatchableError, Exception].}
+  NotificationHandler[T] = proc(input: T): Future[void] {.gcsafe, raises: [Defect, CatchableError, Exception].}
+
+proc extractId  (id: JsonNode): int =
+  if id.kind == JInt:
+    result = id.getInt
+  if id.kind == JString:
+    discard parseInt(id.getStr, result)
+
+proc wrapJsonRpcResponse(s: string): string =
+  result = s & "\r\n"
+  result = "Content-Length: " & $s.len & "\r\n\r\n" & s
+
+proc wrap[T, Q](callback: Handler[T, Q]): RpcProc =
+  return
+    proc(input: JsonNode): Future[RpcResult] {.async} =
+      let res = await callback(to(input{"params"}, T))
+      return some(StringOfJson($(%res)))
+
+proc wrap[T, Q](callback: HandlerWithId[T, Q]): RpcProc =
+  return
+    proc(input: JsonNode): Future[RpcResult] {.async} =
+      let id = input{"id"}.extractId
+      let params = input{"params"}
+      return some(StringOfJson($(%(await callback(to(params, T), id)))))
+
+proc wrap[T](callback: NotificationHandler[T]): RpcProc =
+  return
+    proc(input: JsonNode): Future[RpcResult] {.async} =
+      await callback(to(input{"params"}, T))
+      return none[StringOfJson]()
+
+proc register*[T, Q](server: RpcServer, name: string, rpc: Handler[T, Q]) =
+  server.register(name, wrap(rpc))
+
+proc register*[T, Q](server: RpcServer, name: string, rpc: HandlerWithId[T, Q]) =
+  server.register(name, wrap(rpc))
+
+proc registerNotification*[T](server: RpcServer, name: string, rpc: NotificationHandler[T]) =
+  server.register(name, wrap(rpc))
+
+method call*(self: StreamClient,
+             name: string,
+             params: JsonNode): Future[Response] {.async} =
+  ## Remotely calls the specified RPC method.
+  let
+    id = self.getNextId()
+    value = wrapJsonRpcResponse($rpcCallNode(name, params, id))
+    # completed by processMessage.
+    newFut = newFuture[Response]()
+
+  # add to awaiting responses
+  self.awaiting[id] = newFut
+
+  write(OutputStream(self.output), value)
+  flush(self.output)
+  return await newFut
+
+proc call*(connection: StreamConnection, name: string,
+          params: JsonNode): Future[Response] {.gcsafe, raises: [Exception].} =
+  return connection.client.call(name, params)
+
+proc notify*(connection: StreamConnection, name: string,
+             params: JsonNode): Future[void] {.async.} =
+  let value = wrapJsonRpcResponse($rpcNotificationNode(name, params))
+  write(OutputStream(connection.output), value)
+  flush(connection.output)
+
+proc readMessage*(input: AsyncInputStream): Future[Option[string]] {.async.} =
```

Comment:
async input streams based on faststreams require that all consumers of the new API are based on faststreams/async - this significantly limits the utility of the feature - what should be done  instead is to move the main logic such that it can be used with "pure" chronos without faststreams, then the faststreams support should be implemented as an optional layer on top that can be imported / enabled separately.

Reply:
I just want to point out that `start` is generic, and you can implement `readMessage` for whatever entity you want. I had an implementation over the sync stream(which I dropped before making this PR). Not sure if this will address your concern at 100%.

Reply:
Unfortunately not - the requirement that a specific `-d` flag is used is the problem / non-starter: it "infects" unrelated usage of faststreams globally - thus, it's better to either stick with pure chronos, or develop the faststreams support on the side - this latter option needs work on the faststreams side as well.